imagebuildah/stage_executor.go

```diff
@@ -1168,6 +1194,24 @@ func (s *StageExecutor) Execute(ctx context.Context, base string) (imgID string,
 			if err != nil {
 				return "", nil, fmt.Errorf("error checking if cached image exists from a previous build: %w", err)
 			}
+			// All the best effort to find cache on localstorage have failed try pulling
+			// cache from remote repo if `--cache-from` was configured.
+			if cacheID == "" && s.executor.cacheFrom != "" {
+				src := fmt.Sprintf("%s:%s", s.executor.cacheFrom, cacheKey)
```

Comment:
BTW I don’t think the generic code should be doing this string pasting — just call `pullCache(ctx, cacheKey)`. (`pullCache` itself can read `s.executor.cacheFrom`), and the same for push.

(I wouldn’t even mind a helper function for this one line, shared between `pullCache` and `pushCache`, especially if this one line ends up being a bit longer `docker.NewReference(reference.WithTag(…))` as suggested elsewhere.)

Reply:
BTW `pullCache(ctx, cacheKey)` would nicely almost-isolate the storage backend (registry/S3/…) from the stage executor.

(_almost_-isolate because there is still the `s.executor.cacheFrom != ""` condition which might need abstracting. That can certainly be done as well.)